packages/client/src/__tests__/integration/happy/interactive-transactions-postgres/test.ts

```diff
@@ -0,0 +1,473 @@
+
+      await prisma.user.create({
+        data: {
+          email: 'user_2@website.com',
+        },
+      })
+
+      return prisma.user.findMany()
+    })
+
+    expect(result.length).toBe(2)
+  })
+
+  /**
+   * Transactions should fail after the default timeout
+   */
+  test('timeout default', async () => {
+    const result = prisma.$transaction(async (prisma) => {
+      await prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+
+      await new Promise((res) => setTimeout(res, 6000))
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(
+      `Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.`,
+    )
+
+    expect(await prisma.user.findMany()).toHaveLength(0)
+  })
+
+  /**
+   * Transactions should fail if they time out on `timeout`
+   */
+  test('timeout override', async () => {
+    const result = prisma.$transaction(
+      async (prisma) => {
+        await prisma.user.create({
+          data: {
+            email: 'user_1@website.com',
+          },
+        })
+
+        await new Promise((res) => setTimeout(res, 6000))
+      },
+      {
+        maxWait: 200,
+        timeout: 500,
+      },
+    )
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(
+      `Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.`,
+    )
+
+    expect(await prisma.user.findMany()).toHaveLength(0)
+  })
+
+  /**
+   * Transactions should fail if they time out on `maxWait`
+   */
+  test.skip('maxWait override', async () => {
+    const result = prisma.$transaction(
+      async (prisma) => {
+        await prisma.user.create({
+          data: {
+            email: 'user_1@website.com',
+          },
+        })
+
+        await new Promise((res) => setTimeout(res, 5))
+
+        return prisma.user.findMany()
+      },
+      {
+        maxWait: 0,
+      },
+    )
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+                
+                            Invalid \`prisma.user.findMany()\` invocation:
+                
+                
+                              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.
+                        `)
+  })
+
+  /**
+   * Transactions should fail and rollback if thrown within
+   */
+  test('rollback throw', async () => {
+    const result = prisma.$transaction(async (prisma) => {
+      await prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+
+      throw new Error('you better rollback now')
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`you better rollback now`)
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * A transaction might fail if it's called inside another transaction
```

Comment:
Does it only depend on the database or something else too?

Reply:
I think this comes from the single process nature of sqlite, it will lock the table for the time of the transaction.

Reply:
Ah, so it only fails on SQLite, got it. That makes sense.